controllers/vaultdynamicsecret_controller.go

```diff
@@ -193,6 +189,11 @@ func (r *VaultDynamicSecretReconciler) Reconcile(ctx context.Context, req ctrl.R
 	}
```

Comment:
Do we maybe need to delay revoking the old lease so that consumers have a grace period to load in and start using the newly created lease?

Reply:
I'm going back and forth a little on this one. Maybe the answer is just that people who want overlap shouldn't use the `Revoke` option, as to get to here, the lease has to be expiring relatively soon anyway.

Reply:
Sorry to pop up (I was the one who reported this issue):

> I'm going back and forth a little on this one. Maybe the answer is just that people who want overlap shouldn't use the Revoke option

Maybe I'm missing something (or I'm simply lacking knowledge of the Vault internals), but isn't the overlap given for free by requesting new credentials at 2/3 of the TTL? The revoke of the old credentials will still happen at 3/3 of the TTL, what this change will do is simply ensure that a new credential is generated and available for apps at 2/3 of the TTL. I wouldn't imagine the revoke flow to be changed here. 🤔 The grace period would be 1/3 of the TTL.

Reply:
> I'm going back and forth a little on this one. Maybe the answer is just that people who want overlap shouldn't use the `Revoke` option, as to get to here, the lease has to be expiring relatively soon anyway. Although `Revoke` does also control behaviour on deleting the VDS too, so maybe that's not a perfect answer.

We could perhaps support both. I do like the idea of supporting deferred lease revocation. Definitely something worth considering for the future.

Reply:
I may be misreading the flow of the code as I haven't tested it locally, but it looks to me like every time we hit the line `secretLease, err := r.syncSecret(ctx, vClient, o)` (which generates a new lease), we always then revoke the old lease 2 lines above this comment if `o.spec.Revoke` is set. So from that reading, I would expect the `r.revokeLease(ctx, o, oldLease.ID)` line to be called at 2/3 of the TTL as the code currently stands.

Take it with a pinch of salt though - I would want to test it to be certain.

Reply:
> However, I currently cannot think of a scenario of a dynamic secret that we want to rotate after 2/3 of the TTL (to avoid the issue this PR is addressing,) and at the same time, be willing to revoke the old credential at 2/3 of the TTL, as it would defeat the purpose of renewing earlier.

I agree, having overlap feels like a sensible default. So perhaps `Revoke` should only apply when deleting the VDS. We haven't done a release with `Revoke` yet, so it wouldn't be a breaking change.

> We could perhaps support both. I do like the idea of supporting deferred lease revocation. Definitely something worth considering for the future.

Yeah, deferred lease revocation strikes me as the more advanced use-case so we can save that part for a later date IMO.

Reply:
> I agree, having overlap feels like a sensible default. So perhaps `Revoke` should only apply when deleting the VDS. We haven't done a release with `Revoke` yet, so it wouldn't be a breaking change.

I agree, we should only revoke the lease on delete.

Reply:
I think we can probably do this in a subsequent PR.